src/TypeAdapter/ReflectionTypeAdapter.php

```diff
@@ -262,77 +248,95 @@ public function read(JsonReadable $reader)
     /**
      * Write the value to the writer for the type
      *
-     * @param JsonWritable $writer
      * @param object $value
-     * @return void
+     * @param WriterContext $context
+     * @return array|null
      */
-    public function write(JsonWritable $writer, $value): void
+    public function write($value, WriterContext $context): ?array
     {
         if ($this->skipSerialize || $value === null) {
-            $writer->writeNull();
-            return;
+            return null;
         }
 
         $classExclusionCheck = $this->hasClassSerializationStrategies
-            && (!$this->requireExclusionCheck || ($this->requireExclusionCheck && $this->classMetadata->getAnnotation(ExclusionCheck::class) !== null));
+            && (!$this->requireExclusionCheck || ($this->requireExclusionCheck && $this->classAnnotations->get(ExclusionCheck::class) !== null));
```

Comment:
These complex inlines could maybe be broken into separate functions since they are used in both read and write, and would make the read and write methods easier to read.